### PR TITLE
fixes issue #552 : Scrollable options tabs.

### PR DESCRIPTION
**Pull request by kianzarrin**
_Fixes_ #552 

- put a `UIScrollablePanel` inside each tab page to make it scroll-able
- Encapsulated `UITabstrip` in `ExtUITabstrip`.
- in each MakeSettings* method, used the new `ExtUITabstrip` class and removed duplicated codes.
 - The scroll-able panel does not work with keybinds panel for some reason, so the keybind panels are not scrollable. there is not too many options there anyway. I suggest we deal with it when we need it.
- The panels are not long enough to require scrollbars yet. Not until I merge in #539 in another PR. So you should not notice any visual difference in the appearance yet. 
- Added static Test class to test the scroll-able panels. just uncomment the following to activate the tests:
```public static void MakeSettings(UIHelperBase helper){//ExtUITabstrip.Test.OnSettingsUI(helper);return;```

### CODE REVIEW

**Comment by kvakvs** [comment]
The code so far is looking good.

Please do `git checkout -b 552-scrollable-options`
Then `git push -u` and follow the suggestion (whatever your remote in Git is named, origin probably, in that case `git push -u origin 552-scrollable-options`
And then you can have this work in a branch other than `master`.

**Comment by aubergine10** [comment]
@kianzarrin Will #562 affect this PR? If we plan on switching to new config system (eg. similar to Real Time), it would probably negate some of the work done in this PR?

**Comment by kianzarrin** [comment]
No @aubergine10. #562 does not influence this PR. This is just encapsulation of UI and it runs completely independently of TMPE. Infact you can copy ExtUITabstrip.cs into an empty mod and I will work just fine.

If anything better encapsulation and removing of duplicate code paves the path to any future modification.

In any case this is blocking my other tasks so I need this in ASAP.

File: TLM/TLM/Manager/Impl/ExtUITabStrip.cs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            UITabContainer tabContainer = container.AddUIComponent<UITabContainer>();
+            tabContainer.relativePosition = new Vector3(0, 80);
+            tabContainer.width = tabStrip.width;
+            tabContainer.height = h;
+            tabStrip.tabPages = tabContainer;
+
+            return tabStrip;
+        }
+
+        public static class Test {
```
Comment (aubergine10) [inline TLM/TLM/Manager/Impl/ExtUITabStrip.cs]: This static class should be encapsulated in `#if DEBUG ... #endif` block? I don't see any need for it to be in `RELEASE` or `LABS` builds.

File: TLM/TLM/Manager/Impl/ExtUITabStrip.cs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            scrollablePanel.eventMouseWheel += (component, eventParam) => {
+                verticalScrollbar.value -= (int)eventParam.wheelDelta * verticalScrollbar.incrementAmount;
+            };
+
+            scrollablePanel.verticalScrollbar = verticalScrollbar;
+
+            return verticalScrollbar;
+        }
+
+        private UIScrollablePanel CreateScrollebalePanel(UIPanel panel) {
```
Comment (krzychu124) [inline TLM/TLM/Manager/Impl/ExtUITabStrip.cs]: Typo 😉

File: TLM/TLM/Manager/Impl/ExtUITabStrip.cs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                panelHelper = new UIHelper(scrollablePanel);
+            }
+            else {
+                currentPanel.autoLayoutDirection = LayoutDirection.Vertical;
+                panelHelper = new UIHelper(currentPanel);
+            }
+            return panelHelper;
+        }
+
+        public static ExtUITabstrip Create(UIHelperBase helper_base) {
```
Comment (krzychu124) [inline TLM/TLM/Manager/Impl/ExtUITabStrip.cs]: Follow naming style, don't use underscore in variable names.

File: TLM/TLM/State/Options.cs [hunk 4]
```diff
@@ -115,46 +116,15 @@ public class Options : MonoBehaviour {
...
-            // MAINTENANCE
-            ++tabIndex;
-            OptionsMaintenanceTab.MakeSettings_Maintenance(tabStrip, tabIndex);
-
-            // KEYBOARD
-            ++tabIndex;
-            OptionsKeybindsTab.MakeSettings_Keybinds(tabStrip, tabIndex);
-
-            tabStrip.selectedIndex = 0;
+            //ExtUITabstrip.Test.OnSettingsUI(helper);return;
```
Comment (krzychu124) [inline TLM/TLM/State/Options.cs]: remove or add new preprocessor directive before merge

File: TLM/TLM/Manager/Impl/ExtUITabStrip.cs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+// https://github.com/PropaneDragon/RushHour/blob/release/RushHour/Options/OptionHandler.cs
+// https://github.com/CWMlolzlz/CS-AdvancedToolbar/blob/master/Source/ExpandableToolbar.cs
+
+namespace TrafficManager.Manager.Impl {
+    using UnityEngine;
+    using ICities;
+    using ColossalFramework.UI;
+
+    public sealed class ExtUITabstrip : UITabstrip {
+        private UIScrollbar CreateVerticalScrollbar(UIPanel panel, UIScrollablePanel scrollablePanel) {
```
Comment (krzychu124) [inline TLM/TLM/Manager/Impl/ExtUITabStrip.cs]: If it's possible move scrollbar more to the right edge of options window

File: TLM/TLM/UI/MainMenu/TimedTrafficLightsButton.cs [hunk 6]
```diff
@@ -1,12 +1,12 @@
...
+namespace TrafficManager.UI.MainMenu {
     using State;
 
     public class TimedTrafficLightsButton : MenuToolModeButton {
         protected override ToolMode ToolMode => ToolMode.TimedLightsSelectNode;
 
         protected override ButtonFunction Function => ButtonFunction.TimedTrafficLights;
 
-        public override string Tooltip => Translation.Menu.Get("Tooltip:Timed traffic lights");
+        public override string Tooltip => Translation.Menu.Get("Tooltip:Timed traffic lights") + "\n" + Translation.Menu.Get("Tooltip.Keybinds:Timed traffic lights");
```
Comment (krzychu124) [inline TLM/TLM/UI/MainMenu/TimedTrafficLightsButton.cs]: We don't have that language key yet 😕

**Comment by krzychu124** [comment]
I've noticed that checkboxes on Overlays tab are more dense positioned than before

__Before__
Zrzut ekranu (283)

__This PR__
Zrzut ekranu (282)

File: TLM/TLM/Manager/Impl/ExtUITabStrip.cs (hunk 5 above)
Comment (kianzarrin) [reply TLM/TLM/Manager/Impl/ExtUITabStrip.cs]: Its as right as it can be.

File: TLM/TLM/State/Options.cs (hunk 4 above)
Comment (kianzarrin) [reply TLM/TLM/State/Options.cs]: I will just remove the commented out code. Its only for testing anyway.

File: TLM/TLM/UI/MainMenu/TimedTrafficLightsButton.cs (hunk 6 above)
Comment (kianzarrin) [reply TLM/TLM/UI/MainMenu/TimedTrafficLightsButton.cs]: Don't know how it got here. I got confused. This belongs to another review.

**Comment by aubergine10** [comment]
Not able to test right now, but just before I forget: Ensure that check boxes can be indented. There's one of the tabs where we indent one or more checkbox options.

**Comment by kianzarrin** [comment]
I am ready for the next round of review.

**Comment by kianzarrin** [comment]
Screenshot (100)
The indent works.
(Just in case you are wondering where are the scroll bars, they have autohide if the panel fits the screen)

**Comment by aubergine10** [comment]
One possible thing to check: Should the extended class it be in the `Impl` folder or the `UI` folder?

**Comment by kianzarrin** [comment]
@aubergine10 
> One possible thing to check: Should the extended class it be in the `Impl` folder or the `UI` folder?

Thanks for raising this point. I noticed that all classes that have Ext in them are in manager.impl . but now that I think of it they are also linked by manager factory. 

I have never worked on object oriented programming in a professional context so I have nooo ideaaaa  how to organize these things.

If you want me to move it to somewhere else I'll be more than happy to 😄

**Comment by aubergine10** [comment]
I also am not sure which is best folder. Hopefully @krzychu124 or @kvakvs can advise.

**Comment by kvakvs** [comment]
I don't have a strong opinion on this. Don't stress yourself too much over file location.
Use something that looks right today, moving it later is a matter of two clicks.

**Comment by kianzarrin** [comment]
As the commit messages suggests, in order to make the changes easier to review  I moved ExtUITabStrip to UI.Helpers folder and namespace in one commit and changed nothing else. Then I fixed the padding issue in separate commit. 

@krzychu124 
> I've noticed that checkboxes on Overlays tab are more dense positioned than before

fixed it. 
Screenshot (108)

**Pull request merged**